### PR TITLE
Exp manager log rank 0 only arguments

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Adds log_local_rank_0_only and log_global_rank_0_only args to exp_manager.

**Collection**: All

# Changelog 
- Makes creation of log files based on world size configurable instead of hard-coded.

# Usage
* exp_manager.log_rank_0_only: True
* exp_manager.log_global_rank_0_only: False

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/utils/exp_manager.py [hunk 1]
```diff
@@ -216,7 +219,10 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
                 Defaults to True.
             - files_to_copy (list): A list of files to copy to the experiment logging directory. Defaults to None which
                 copies no files.
-
+            - log_local_rank_0_only (bool): Whether to only log create log files only for local rank 0. Defaults to False.
```
Comment (ericharper) [inline nemo/utils/exp_manager.py:222]: Typo: log create log -> create log

**Comment by mlgill** [comment]
@MaximumEntropy I can't mark myself as a reviewer, probably since I'm not a member of this repo, but I reviewed and it looks good. Thanks!

File: nemo/utils/exp_manager.py [hunk 2]
```diff
@@ -216,7 +219,10 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
...
+                Set this to True if you are using DDP with many GPUs and do not want many log files in your exp dir.
+            - log_global_rank_0_only (bool): Whether to only log create log files only for global rank 0. Defaults to False.
```
Comment (ericharper) [inline nemo/utils/exp_manager.py:224]: Typo: log create log -> create log

File: nemo/utils/exp_manager.py [hunk 3]
```diff
@@ -281,16 +287,25 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
...
+            f"Cannot set both log_local_rank_0_only and log_global_rank_0_only to True. Please set either one or neither."
+        )
+
     # Handle logging to file
-    if get_envbool(NEMO_ENV_VARNAME_TESTING, False) or world_size <= 32:
+    # Logs on all ranks.
+    if get_envbool(NEMO_ENV_VARNAME_TESTING, False) or (
+        not cfg.log_local_rank_0_only and not cfg.log_global_rank_0_only
+    ):
         # If NEMO_TESTING is set (debug mode) or if less than 32 ranks save all log files
```
Comment (ericharper) [inline nemo/utils/exp_manager.py:300]: This comment should remove the `or if less than 32 ranks save all log files`

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

File: nemo/utils/exp_manager.py [hunk 4]
```diff
@@ -281,17 +287,25 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
...
+        raise ValueError(
+            f"Cannot set both log_local_rank_0_only and log_global_rank_0_only to True. Please set either one or neither."
+        )
+
     # Handle logging to file
-    if get_envbool(NEMO_ENV_VARNAME_TESTING, False) or world_size <= 32:
-        # If NEMO_TESTING is set (debug mode) or if less than 32 ranks save all log files
+    # Logs on all ranks.
+    if get_envbool(NEMO_ENV_VARNAME_TESTING, False) or (
+        not cfg.log_local_rank_0_only and not cfg.log_global_rank_0_only
```
Comment (titu1994) [inline nemo/utils/exp_manager.py:298]: Instead of this not, perform explicit cfg.X is True and cfg.Y is True

File: nemo/utils/exp_manager.py [hunk 5]
```diff
@@ -281,17 +287,25 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
     logging.info(f'Experiments will be logged at {log_dir}')
     trainer._default_root_dir = log_dir
 
+    if cfg.log_local_rank_0_only and cfg.log_global_rank_0_only:
+        raise ValueError(
+            f"Cannot set both log_local_rank_0_only and log_global_rank_0_only to True. Please set either one or neither."
```
Comment (titu1994) [inline nemo/utils/exp_manager.py:292]: Why can't both be set - log global rank 0 and all local rank 0s (duplicate global 0 and local 0 if that's easier) then?

**Review by titu1994 — CHANGES_REQUESTED** [review]
Forgot to request changes for the if check

File: nemo/utils/exp_manager.py (hunk 5 above)
Comment (MaximumEntropy) [reply nemo/utils/exp_manager.py:292]: But thats the same behaviour as just log_local_rank_0_only no?

Comment (MaximumEntropy) [reply nemo/utils/exp_manager.py:292]: lets take an example, if using 4 nodes with 16 GPUs.

When setting `log_local_rank_0_only`, you'll end up with 4 log files `nemo_log_globalrank-0_localrank-0.txt`, `nemo_log_globalrank-16_localrank-0.txt`, `nemo_log_globalrank-32_localrank-0.txt`, `nemo_log_globalrank-48_localrank-0.txt`

With `log_global_rank_0_only`, you'll get just one file `nemo_log_globalrank-0_localrank-0.txt`, which is already a part of `log_local_rank_0_only`.

File: nemo/utils/exp_manager.py (hunk 4 above)
Comment (MaximumEntropy) [reply nemo/utils/exp_manager.py]: Fixed.

File: nemo/utils/exp_manager.py (hunk 5 above)
Comment (titu1994) [reply nemo/utils/exp_manager.py:292]: Ok so local 0 global 0 anyway is global 0, that's fine to make them exclusive flags then.

File: nemo/utils/exp_manager.py [hunk 6]
```diff
@@ -281,17 +287,25 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
...
+        )
+
+    # This is set if the env var NEMO_TESTING is set to True.
+    nemo_testing = get_envbool(NEMO_ENV_VARNAME_TESTING, False)
+
     # Handle logging to file
-    if get_envbool(NEMO_ENV_VARNAME_TESTING, False) or world_size <= 32:
-        # If NEMO_TESTING is set (debug mode) or if less than 32 ranks save all log files
+    # Logs local rank 0 only
+    if local_rank == 0 and cfg.log_local_rank_0_only and not nemo_testing:
```
Comment (titu1994) [inline nemo/utils/exp_manager.py:300]: Dont use implicit truths like this. Yes we know the flag is true or false but don't check it like this. Just use cfg.log_local_... is True/False as you need it here.

**Review by titu1994 — APPROVED** [review]
LGTM thanks

**Pull request merged**